lib/rust/ensogl/pack/src/assets.rs

```diff
@@ -0,0 +1,362 @@
+        for entry in builder_dir {
+            let entry = entry?;
+            let asset_key = entry.file_name().to_string_lossy().to_string();
+            let dir = ide_ci::fs::read_dir(entry.path())?;
+            let mut file_hashes = BTreeMap::new();
+            for entry in dir {
+                let entry = entry?;
+                let file_name = entry.file_name().to_string_lossy().to_string();
+                let path = entry.path();
+                buf.clear();
+                ide_ci::fs::open(path)?.read_to_end(&mut buf)?;
+                let mut file_hasher = std::collections::hash_map::DefaultHasher::new();
+                buf.hash(&mut file_hasher);
+                file_hashes.insert(file_name, file_hasher.finish());
+            }
+            let mut asset_hasher = std::collections::hash_map::DefaultHasher::new();
+            file_hashes.hash(&mut asset_hasher);
+            let inputs_hash = asset_hasher.finish();
+            let input_files = file_hashes.into_keys().collect();
+            builder_sources.push(AssetSources { asset_key, input_files, inputs_hash });
+        }
+    }
+    Ok(asset_sources)
+}
+
+/// Generate any assets not found up-to-date in the cache.
+async fn update_assets(
+    paths: &Paths,
+    sources: &HashMap<Builder, Vec<AssetSources>>,
+) -> Result<AssetManifest> {
+    let out = &paths.target.ensogl_pack.dist.dynamic_assets;
+    ide_ci::fs::create_dir_if_missing(out)?;
+    let mut assets: AssetManifest = BTreeMap::new();
+    let mut deferred_assets: BTreeMap<Builder, Vec<_>> = BTreeMap::new();
+    for (&builder, builder_sources) in sources {
+        let out = out.join(builder.dir_name());
+        ide_ci::fs::create_dir_if_missing(&out)?;
+        for source_specification in builder_sources {
+            let out = out.join(source_specification.dir_name());
+            let key = source_specification.asset_key.clone();
+            match std::fs::try_exists(&out)? {
+                false => {
+                    info!("Rebuilding asset: `{}`.", out.display());
+                    let builder_assets = deferred_assets.entry(builder).or_default();
+                    let build = build_asset(paths, builder, source_specification);
+                    builder_assets.push(async move { Ok((key, build.await?)) });
+                }
+                true => {
+                    debug!("Skipping clean asset: `{}`.", out.display());
+                    let builder_assets = assets.entry(builder).or_default();
+                    let asset = survey_asset(paths, builder, source_specification)?;
+                    builder_assets.insert(key, asset);
+                }
+            };
+        }
+    }
+    for (builder, deferred_assets) in deferred_assets.into_iter() {
+        let deferred_assets = futures::future::join_all(deferred_assets).await;
+        let deferred_assets: Result<Vec<_>> = deferred_assets.into_iter().collect();
```

Comment:
`try_collect_vec` could be helpful here.

Reply:
It breaks type inference badly here, because its output error type is invariant in its input error parameter (`E: Into<anyhow::Error>`).